fix(wallet, wallet-dashboard): show inactive validator metadata

KeitoTadashi opened the pull request:
# Description of change



## Links to any relevant issues

fixes #6480 

## Type of change

Choose a type of change, and delete any options that are not relevant.

- Bug fix (a non-breaking change which fixes an issue)
- Enhancement (a non-breaking change which adds functionality)
- Breaking change (fix or feature that would cause existing functionality to not work as expected)
- Documentation Fix

## How the change has been tested

Describe the tests that you ran to verify your changes.

Make sure to provide instructions for the maintainer as well as any relevant configurations.

## Change checklist

Tick the boxes that are relevant to your changes, and delete any items that are not.

- [ ] I have followed the contribution guidelines for this project
- [ ] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have checked that new and existing unit tests pass locally with my changes

### Release Notes

<!--
Check each box that your changes affect. If none of the boxes relate to your changes, release notes aren't required.

For each box you select, include information after the relevant heading that describes the impact of your changes that a user might notice and any actions they must take to implement updates.
-->

- [ ] Protocol:
- [ ] Nodes (Validators and Full nodes):
- [ ] Indexer:
- [ ] JSON-RPC:
- [ ] GraphQL:
- [ ] CLI:
- [ ] Rust SDK:
- [ ] REST API:

KeitoTadashi committed:
fix(wallet-dashboard): show inactive validator metadata

KeitoTadashi committed:
fix(wallet-dashboard): remove logs

KeitoTadashi committed:
Merge branch 'develop' into tooling/fetch-inactive-validators-for-wallet-and-walletdashboard

KeitoTadashi committed:
fix(wallet-dashboard): fix getLatestIotaSystemState

KeitoTadashi committed:
Merge branch 'develop' into tooling/fetch-inactive-validators-for-wallet-and-walletdashboard

brancoder commented on apps/core/src/utils/stake/getInactiveValidatorsData.ts:
Should we move these client calls to some hook instead of utils method. Non of the utils deal with fetching the data from the nodes.

brancoder commented on apps/core/src/constants/schema.constants.ts:
We should move the schemas to a better placed types file, because these constant files usually contain only constants

KeitoTadashi committed:
Merge branch 'develop' into tooling/fetch-inactive-validators-for-wallet-and-walletdashboard

KeitoTadashi committed:
fix((wallet-dashboard): relocate files

KeitoTadashi committed:
fix(wallet-dashboard): update imports

KeitoTadashi committed:
fix(wallet-dashboard): update imports

marc2332 commented on apps/core/src/components/stake/StakedCard.tsx:
```diff
@@ -49,17 +51,50 @@ export function StakedCard({
+        },
+    });
+
+    const combinedValidatorData = useMemo(() => {
+        if (!validatorMeta && !inactiveValidatorData) return null;
+        return {
+            ...validatorMeta,
+            ...inactiveValidatorData,
+        };
+    }, [validatorMeta, inactiveValidatorData]);
```
Don't you maybe want to do something like this?

(No need for `useMemo` this is a cheap operation btw)

```suggestion
    const validatorData = validatorMeta || inactiveValidatorData || null;
```

marc2332 commented on apps/core/src/hooks/stake/useGetInactiveValidatorsData.ts:
This isn't a hook 👀

KeitoTadashi replied:
Maybe should i place in utils?

marc2332 commented on apps/core/src/hooks/stake/useValidatorInfo.ts:
```diff
@@ -10,13 +15,48 @@ export function useValidatorInfo({ validatorAddress }: { validatorAddress: strin
         system?.activeValidators.find((validator) => validator.iotaAddress === validatorAddress) ||
         null;
+    const { data: inactiveValidatorData } = useQuery({
+        queryKey: [system?.inactivePoolsId, validatorAddress],
+        async queryFn() {
+            if (!system?.inactivePoolsId || !validatorAddress) {
+                throw Error('Missing params');
+            }
+            const inactiveValidators = await iotaClient.getDynamicFields({
+                parentId: normalizeIotaAddress(system?.inactivePoolsId),
+            });
 
-    const currentEpoch = Number(system?.epoch || 0);
+            const pendingInactiveValidatorsData = await Promise.all(
+                inactiveValidators.data.map(
+                    async (validator) =>
+                        await getInactiveValidatorsData(iotaClient, validator.objectId),
+                ),
+            );
 
+            return pendingInactiveValidatorsData;
+        },
+        enabled: !!system?.inactivePoolsId && !!validatorAddress,
+        select(validators) {
+            return validators.find((validator) => validator?.validatorAddress === validatorAddress);
+        },
+    });
```
Would be nice if we can avoid duplicating logic, can't we maybe move the inactive data fetching of validators to a separate hook?

KeitoTadashi committed:
Merge branch 'develop' into tooling/fetch-inactive-validators-for-wallet-and-walletdashboard

KeitoTadashi committed:
fix(wallet-dashboard): remove duplicate code and relocate hooks

KeitoTadashi committed:
fix(wallet-dashboard): update imports

KeitoTadashi committed:
Merge branch 'develop' into tooling/fetch-inactive-validators-for-wallet-and-walletdashboard

KeitoTadashi committed:
Merge branch 'develop' into tooling/fetch-inactive-validators-for-wallet-and-walletdashboard

KeitoTadashi committed:
Merge branch 'develop' into tooling/fetch-inactive-validators-for-wallet-and-walletdashboard

marc2332 commented:
Left some suggestions in https://github.com/iotaledger/iota/compare/tooling/fetch-inactive-validators-for-wallet-and-walletdashboard...tooling/fetch-inactive-validators-for-wallet-and-walletdashboard-marc

KeitoTadashi committed:
fix(wallet/dashboard/explorer): get inactive validators

KeitoTadashi committed:
Merge branch 'develop' into tooling/fetch-inactive-validators-for-wallet-and-walletdashboard

marc2332 commented on apps/core/src/hooks/stake/useValidatorInfo.ts:
I dont see we we need to call `useGetInactiveValidators` here, I just call call the hook in the components where its needed

KeitoTadashi replied:
its needed cause i added inactiveValidatorSummary in return

marc2332 replied:
I know,but what I am saying is that `inactiveValidatorSummary` should not be returned by that hook

marc2332 commented on apps/core/src/hooks/stake/useGetInactiveValidator.ts:
```diff
@@ -0,0 +1,44 @@
+    const iotaClient = useIotaClient();
+    const { data } = useIotaClientQuery('getLatestIotaSystemState');
+    const inactivePoolsId = data?.inactivePoolsId;
+    const queryResult = useQuery({
+        queryKey: [inactivePoolsId, validatorAddress],
+        async queryFn() {
+            if (!inactivePoolsId || !validatorAddress) {
+                throw Error('Missing params');
+            }
+            const inactiveValidators = await iotaClient.getDynamicFields({
+                parentId: normalizeIotaAddress(inactivePoolsId),
+            });
+            const pendingInactiveValidatorsData = await Promise.all(
+                inactiveValidators.data.map(
+                    async (validator) =>
+                        await getInactiveValidatorsMetadata(iotaClient, validator.objectId),
+                ),
+            );
+            return pendingInactiveValidatorsData;
+        },
+        enabled: !!inactivePoolsId && !!validatorAddress,
+    });
+
+    if (queryResult.isLoading || queryResult.isError || !queryResult.data) {
+        return null;
+    }
+
+    const validatorData = queryResult.data.find(
+        (validator) => validator?.validatorAddress === validatorAddress,
+    );
+
+    return validatorData || null;
+}
```
I would rather make the whole query available to the caller of the hook:
```suggestion
    return useQuery({
        queryKey: [inactivePoolsId, validatorAddress],
        async queryFn() {
            if (!inactivePoolsId || !validatorAddress) {
                throw Error('Missing params');
            }
            const inactiveValidators = await iotaClient.getDynamicFields({
                parentId: normalizeIotaAddress(inactivePoolsId),
            });
            const pendingInactiveValidatorsData = await Promise.all(
                inactiveValidators.data.map(
                    async (validator) =>
                        await getInactiveValidatorsMetadata(iotaClient, validator.objectId),
                ),
            );
            return pendingInactiveValidatorsData.find(
		        (validator) => validator?.validatorAddress === validatorAddress,
		    );
        },
        enabled: !!inactivePoolsId && !!validatorAddress,
    });
}
```

KeitoTadashi committed:
Merge branch 'develop' into tooling/fetch-inactive-validators-for-wallet-and-walletdashboard

KeitoTadashi committed:
Merge branch 'develop' into tooling/fetch-inactive-validators-for-wallet-and-walletdashboard

msarcev committed:
feat(explorer): Add next epoch commission column to Validators (#6809)

## Links to any relevant issues

Closes https://github.com/iotaledger/iota/issues/6808

Co-authored-by: Bran <52735957+brancoder@users.noreply.github.com>

msarcev committed:
fix(explorer): Render active validator data over inactive if available (#6813)

On `/validator` page, if both activeValidator data and inactive is
found, we render the active data.

## Links to any relevant issues

Closes https://github.com/iotaledger/iota/issues/6812

KeitoTadashi committed:
fix(wallet, dashboard, explorer): improve inactive validators

KeitoTadashi committed:
fix typo

KeitoTadashi committed:
Merge branch 'develop' into tooling/fetch-inactive-validators-for-wallet-and-walletdashboard

marc2332 commented on apps/core/src/hooks/stake/useValidatorInfo.ts:
```diff
@@ -10,13 +12,35 @@ export function useValidatorInfo({ validatorAddress }: { validatorAddress: strin
+                validatorStakingPoolId: inactiveValidatorData?.validatorStakingPoolId || '',
+            };
+        }
+    }
```
Let's remove this from here

https://github.com/iotaledger/iota/pull/6550#discussion_r2081054023

Its not like the hook can be dynamically called anyway

KeitoTadashi committed:
fix(wallet, dashboard): remove inactiveValidator from useValidatorInfo

marc2332 commented:
Conflicts @KeitoTadashi

KeitoTadashi committed:
Merge branch 'develop' into tooling/fetch-inactive-validators-for-wallet-and-walletdashboard

brancoder commented on apps/core/src/components/stake/StakedCard.tsx:
```diff
@@ -59,16 +58,16 @@ export function StakedCard({
             </CardImage>
             <CardBody
-                title={name}
+                title={validatorData?.name || '--'}
```
we dont need to change this line, we can use `title={name}` here

brancoder commented on apps/core/src/components/stake/StakedCard.tsx:
```diff
@@ -59,16 +58,16 @@ export function StakedCard({
     return (
         <Card testId="staked-card" type={CardType.Default} isHoverable onClick={onClick}>
             <CardImage>
                 <ImageIcon src={validatorMeta?.imageUrl || null} label={name} fallback={name} />
```
we should use `validatorData?.imageUrl` instead of `validatorMeta?.imageUrl

KeitoTadashi replied:
yes u are right!

brancoder commented on apps/wallet-dashboard/components/dialogs/staking/views/DetailsView.tsx:
```diff
@@ -75,7 +76,9 @@ export function DetailsView({
 
-    const validatorName = validatorSummary?.name || '--';
+    const { data: inactiveValidatorSummary } = useGetInactiveValidator(validatorAddress);
+    const validatorName = inactiveValidatorSummary?.name || validatorSummary?.name;
```
we should fallback to validator address so the code is in line with other similar components like StakedCard
`    const validatorName = inactiveValidatorSummary?.name || validatorSummary?.name || validatorAddress;
`

brancoder reviewed (changes requested):
LG just few minor changes

KeitoTadashi committed:
fix(wallet, dashboard): requested changes

KeitoTadashi committed:
Merge branch 'develop' into tooling/fetch-inactive-validators-for-wallet-and-walletdashboard

brancoder reviewed (changes requested):
LG and works great we just need to polish the `useGetInactiveValidator` because when there is NO inactive validators, and user has active stake to a valid committee member, this error is logged in
![image](https://github.com/user-attachments/assets/3706ca23-82b2-41d6-a3a8-3f2ab1add975)

KeitoTadashi committed:
fix(wallet, dashboard): manage when no inactiveValidators

KeitoTadashi committed:
Merge branch 'develop' into tooling/fetch-inactive-validators-for-wallet-and-walletdashboard

brancoder reviewed (changes requested):
The undefined  query result is still being shown on Select a Validator page
<img width="1692" alt="Screenshot 2025-05-15 at 08 36 03" src="https://github.com/user-attachments/assets/14d49187-7147-49b0-a92d-fb62f3460e7a" />

KeitoTadashi committed:
fix(wallet, dashboard): fix hook

KeitoTadashi committed:
Merge branch 'develop' into tooling/fetch-inactive-validators-for-wallet-and-walletdashboard

brancoder commented on apps/core/src/hooks/stake/useGetInactiveValidator.ts:
```diff
@@ -30,9 +30,13 @@ export function useGetInactiveValidator(validatorAddress: string) {
+                if (!findValidator) {
+                    return null;
+                }
+                return findValidator;
```
can we simplify this code to
```
const validator = pendingInactiveValidatorsData.find(
                (validator) => validator?.validatorAddress === validatorAddress,
            ) as InactiveValidatorData;

            return validator ?? null;
```

KeitoTadashi committed:
fix(wallet, dashboard): fix requested changes

KeitoTadashi committed:
Merge branch 'develop' into tooling/fetch-inactive-validators-for-wallet-and-walletdashboard

brancoder reviewed (approved):
LGTM nice job 👍

VmMad commented on apps/core/src/hooks/stake/useGetInactiveValidator.ts:
```diff
@@ -0,0 +1,37 @@
+            );
+            const validator = pendingInactiveValidatorsData.find(
+                (validator) => validator?.validatorAddress === validatorAddress,
+            ) as InactiveValidatorData;
```
```suggestion
            );
```

VmMad commented on apps/core/src/utils/stake/getInactiveValidatorsMetadata.ts:
```diff
@@ -0,0 +1,54 @@
+import { InactiveValidatorData, ValidatorSchema, DynamicFieldObjectSchema } from '../../types';
+
+// Function to get inactive validator data
+// It fetches the validator object and its dynamic fields to extract metadata
```
I think the fn name is self explaining

VmMad commented on apps/core/src/utils/stake/getInactiveValidatorsMetadata.ts:
```diff
@@ -0,0 +1,54 @@
+// It fetches the validator object and its dynamic fields to extract metadata
+export async function getInactiveValidatorsMetadata(
+    client: IotaClient,
+    objectId: string,
```
```suggestion
    validatorObjectId: string,
```

VmMad commented on apps/core/src/utils/stake/getInactiveValidatorsMetadata.ts:
```diff
@@ -0,0 +1,54 @@
+            ),
+        ),
+        validatorStakingPoolId: validator.data.fields.name,
+    };
```
```suggestion

    const { image_url, description, name, project_url, iota_address, protocol_pubkey_bytes } =
        metadata.data.fields.value.fields.metadata.fields;

    return {
        imageUrl: image_url,
        description: description,
        name: name,
        projectUrl: project_url,
        validatorAddress: iota_address,
        validatorPublicKey: toB64(Uint8Array.from(protocol_pubkey_bytes)),
        validatorStakingPoolId: validator.data.fields.name,
    };
```

VmMad replied:
or another way just assign a constant
```tsx
const validatorMetadata = metadata.data.fields.value.fields.metadata;
return {
    imageUrl: validatorMetadata.image_url,
    // ...
```

VmMad commented on apps/wallet-dashboard/components/dialogs/staking/views/DetailsView.tsx:
```diff
@@ -77,7 +78,11 @@ export function DetailsView({
 
-    const validatorName = validatorSummary?.name || '--';
+    const { data: inactiveValidatorSummary } = useGetInactiveValidator(validatorAddress);
+    const validatorName =
+        inactiveValidatorSummary?.name || validatorSummary?.name || validatorAddress;
+    const validatorImageUrl =
+        inactiveValidatorSummary?.imageUrl || validatorSummary?.imageUrl || null;
```
why use the inactiveValidatorSummary first instead of using the normal one and fallback to inactive?

KeitoTadashi committed:
fix(dashboard): requested changes

brancoder reviewed (approved):
LGTM

msarcev commented on apps/core/src/utils/stake/getInactiveValidatorsMetadata.ts:
```diff
@@ -0,0 +1,50 @@
+        },
+    });
+    const metadata = DynamicFieldObjectSchema.safeParse(dfObject.data?.content);
+    if (!metadata.data || !validator.data) {
```
I'm aware that this code was moved here but it seams to me that this `!validator.data` is redundant here. Isn't this condition already covered on  L19 and L20 ? 🤔 Should we remove it ?

marc2332 replied:
Indeed, let's remove it

msarcev reviewed (commented):
LGTM apart from this small comment. Nice job!

KeitoTadashi committed:
Merge branch 'develop' into tooling/fetch-inactive-validators-for-wallet-and-walletdashboard

msarcev reviewed (changes requested):
Let's remove the redundant condition in `getInactiveValidatorsMetadata.ts` and then we can merge

KeitoTadashi committed:
fix(core): suggested changes

KeitoTadashi committed:
Merge branch 'develop' into tooling/fetch-inactive-validators-for-wallet-and-walletdashboard

KeitoTadashi committed:
revert suggested changes

KeitoTadashi committed:
Merge branch 'develop' into tooling/fetch-inactive-validators-for-wallet-and-walletdashboard

KeitoTadashi committed:
conflicts

evavirseda committed:
Merge branch 'develop' into tooling/fetch-inactive-validators-for-wallet-and-walletdashboard

KeitoTadashi committed:
Merge branch 'develop' into tooling/fetch-inactive-validators-for-wallet-and-walletdashboard

msarcev commented:
Closed in favour of this "clone" PR https://github.com/iotaledger/iota/pull/6924
